### ISSUE TITLE
Multi-GPU training hangs sometimes when fine tuning the RTMDet instance segmentation model

### ISSUE THREAD

**Issue by JKelle** [open-mmlab/mmdetection#10493]
Thanks for your error report and we appreciate it a lot.

**Checklist**

[X] I have searched related issues but cannot get the expected help.
[] I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help. <-- this link is broken for me
[X] The bug has not been fixed in the latest version.

**Describe the bug**
Multi-GPU training hangs sometimes when fine tuning the RTMDet instance segmentation model (based on `configs/rtmdet/rtmdet-ins_s_8xb32-300e_coco.py`). 

I'm training on 1 machine with 4 GPUs, using the pytorch launcher. During the backward pass, all 4 the GPUs go to 100% utilization and stay there. The 4 CPU processes also go to 100% and stay there.

The problem goes away when I switch to detection instead of instance segmentation, which suggests to me the problem is related to the mask head, but I'm not 100% sure. The forward pass seems to work fine. I am able to print out the loss before backward is called. The hanging happens during the backward pass.

The problem also goes away when I use 1 GPU (still using pytorch's distributed `elastic_launch`, just setting `nproc_per_node = 1`).

Depending on which dataset I train on, the problem may or may not happen. I haven't been able to identify why certain data samples cause the problem. I have created a minimal, reproducible example with a custom dataset of 2 images that immediately causes the training to hang. 

I've experienced this issue both with torch version `1.10.0` and `2.0.1`.

**Reproduction**

1. What command or script did you run?
Download the following zip file: [reproduce_training_issue.zip](https://github.com/open-mmlab/mmdetection/files/11728238/reproduce_training_issue.zip)
Unzip it. It should contain 3 files: 

a. `demo_main.py` contains the training logic.
b. `demo_config.py` contains the whole mmdetection Config used.
c. `dataset.pkl` is a pickle file containing a list of 2 data samples already formatted for model consumption.

Run the following to reproduce the issue. Make sure your machine has 4 GPUs. (The problem can be reproduced with as few as 2 GPUs, just change the argument in demo_main.py if you want to use a different number of GPUs).
```
python demo_main.py
```

2. Did you make any modifications on the code or config? Did you understand what you have modified?
Yes, all my modifications are in the zip file.

3. What dataset did you use?
I've provided a custom dataset in the zip file.

**Environment**

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment information and paste it here.
```
sys.platform: linux
Python: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:01:55) [GCC 11.3.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0,1,2,3: Tesla T4
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.7, V11.7.99
GCC: gcc (conda-forge gcc 12.2.0-19) 12.2.0
PyTorch: 1.10.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2022.1-Product Build 20220311 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX512
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.11.1
OpenCV: 4.5.5
MMEngine: 0.7.2
MMDetection: 3.0.0+ecac3a7
```

2. You may add addition that may be helpful for locating the problem, such as
   - I'm using a conda environment.
 ```
$ conda list | grep torch
ffmpeg                    4.3                  hf484d3e_0    pytorch
pytorch                   1.10.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
pytorch-mutex             1.0                        cuda    pytorch
torchvision               0.11.1               py38_cu113    pytorch
```
```
$ echo $PATH
/opt/conda/envs/mmyolo-jkelle/bin:/opt/conda/condabin:/opt/amazon/openmpi/bin:/opt/amazon/efa/bin:/usr/local/cuda/bin:/usr/local/cuda/include:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
```
$ echo $LD_LIBRARY_PATH
/opt/amazon/efa/lib:/opt/amazon/openmpi/lib:/usr/local/cuda/efa/lib:/usr/local/cuda/lib:/usr/local/cuda:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/targets/x86_64-linux/lib:/usr/local/lib:/usr/lib
```
My $PYTHONPATH is empty

**Comment by michaellee1** [comment]
Hi @Czm369 @hhaAndroid,

Is there any answer for why this is happening?  I'm having the exact same problem.
